compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -153,6 +153,13 @@ impl<'thir, 'tcx> Visitor<'thir, 'tcx> for UnsafetyVisitor<'tcx> {
                     self.requires_unsafe(expr.span, CallToUnsafeFunction);
                 }
             }
+            ExprKind::Field { lhs: Expr { ty, .. }, .. } => {
+                if let ty::Adt(adt_def, _) = ty.kind() {
+                    if adt_def.is_union() {
+                        self.requires_unsafe(expr.span, AccessToUnionField);
+                    }
+                }
```

Comment:
Hm, the actual check seems quite a bit more complicated! (cc @RalfJung, who authored those changes)

https://github.com/rust-lang/rust/blob/24379879acc0959e8ae0f3c19d249b3beb278836/compiler/rustc_mir/src/transform/check_unsafety.rs#L214-L266

I'm wondering why more tests didn't fail. I think we're going to have to tweak these rules here.

Reply:
Oh, right, you don't need `unsafe` to *write* to the union. (Tbh I've never used unions in Rust :smile:)

> I'm wondering why more tests didn't fail.

I think we don't run the existing check-fail tests, see my comment above.

Reply:
*All* reads and *some* writes should be `unsafe`, yes -- and we should have tests covering that... tough probably not very precise ones.

Reply:
Yes, the `src/test/ui/union/union-unsafe.rs` should help; you probably want to run that under thirunsafeck as well.

Reply:
@RalfJung I'm trying to properly check if something is reading from a union, and encountered this weirdness. Is this a bug or feature of the MIR borrow checker?

```rust
// this is safe
match foo {
    Foo { zst: () } => {},
}
// but this is so unsafe it prints the same error message twice?
match foo {
    Foo { pizza: Pizza { topping: Some(PizzaTopping::Cheese) | Some(PizzaTopping::Pineapple) | None } } => {},
    // ^~ ERROR access to union field is unsafe and requires unsafe function or block
}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=b0357dbf1b7251ca0842f9215f3381d5)

Reply:
I think that behavior in the first match is a bug in the MIR borrowck, since [RFC 1444](https://rust-lang.github.io/rfcs/1444-union.html#pattern-matching) says that pattern matching against unions must be unsafe. (and MIR borrowck is okay with it since it doesn't actually involve any reads)

Reply:
I guess this has to do with how pattern matching is compiled -- matching `zst` against the `()` pattern does not actually read anything. So it makes sense to me that the former would not be `unsafe`. But one could easily decide either way and both ways make sense.

Reply:
Note that this pattern *is* `unsafe`:
```rust
Foo { zst: x }
```

Reply:
I've updated the checking, I *think* it should properly detect if something is a read or write to union now (I am also now running almost all of the `src/test/ui/union` test suite).